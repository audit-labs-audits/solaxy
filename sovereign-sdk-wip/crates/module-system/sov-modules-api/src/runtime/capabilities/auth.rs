//! This module defines abstractions and workflows around authenticating and authorizing
//! transactions within a rollup.
use std::marker::PhantomData;

use borsh::{BorshDeserialize, BorshSerialize};
use serde::{Deserialize, Serialize};
use sov_modules_macros::config_value_private;
use sov_rollup_interface::crypto::CredentialId;
use sov_rollup_interface::da::DaSpec;
use sov_rollup_interface::TxHash;
use sov_state::User;
use thiserror::Error;

use crate::transaction::{
    AuthenticatedTransactionAndRawHash, Credentials, Transaction, TransactionVerificationError,
    VersionedTx,
};
use crate::{
    capabilities, metered_credential, AuthenticatedTransactionData, Context, CryptoSpec,
    DispatchCall, FullyBakedTx, GasMeter, GasMeteringError, MeteredBorshDeserialize,
    MeteredBorshDeserializeError, MeteredHasher, ProvableStateReader, RawTx, Runtime, Spec,
    StateAccessor,
};

/// The chain ID of the rollup.
pub fn config_chain_id() -> u64 {
    config_value_private!("CHAIN_ID")
}

/// A batch sent by an unregistered sequencer contains only one transaction.
pub struct BatchFromUnregisteredSequencer {
    /// The transaction.
    pub tx: RawTx,
    /// Id of the batch.
    pub id: [u8; 32],
}

/// Authenticates raw transactions, ensuring that the *claimed* sender really did sign off on the transaction. Note that
/// simply *authenticating* a transaction does not guarantee that it will actually be executed. That decision is
/// made by the [`TransactionAuthorizer`]
///
/// Implementations of this trait should provide a way to interpret the raw bytes of the transaction and authenticate it.
/// Typically, the authentication will require checking the signature of the transaction.
pub trait TransactionAuthenticator<S: Spec> {
    /// The "message" that is extracted from the transaction and passed to the runtime for execution.
    type Decodable;

    /// The input to the authenticator
    type Input: BorshDeserialize + BorshSerialize + Clone + std::fmt::Debug + Send + Sync + 'static;

    /// Authenticates a transaction (typically by checking the signature) and deserializes its contents
    /// into an executable message.
    fn authenticate<Accessor: ProvableStateReader<User, Spec = S>>(
        tx: &FullyBakedTx,
        state: &mut Accessor,
    ) -> Result<AuthenticationOutput<S, Self::Decodable>, AuthenticationError>;

    /// MUST return the same hash as [`TransactionAuthenticator::authenticate`].
    #[cfg(feature = "native")]
    fn compute_tx_hash(tx: &FullyBakedTx) -> anyhow::Result<TxHash>;

    /// Decode a transaction into a message.
    /// This method doesn’t charge gas for deserialization, so it’s meant for off-chain code only (hence to the `native` feature).
    #[cfg(feature = "native")]
    fn decode_serialized_tx(tx: &FullyBakedTx) -> Result<Self::Decodable, FatalError>;

    /// Authenticates raw transaction that is submitted from unregistered sequencers for the
    /// purpose of forced registration (circumventing censorship by currently registered sequencers).
    ///
    /// Implementers of this method should take care to ensure that the gas consumption of this method is low because
    /// (if authentication fails) no one pays for the gas consumed by the authentication check.
    ///
    /// This is *not*  a significant DOS vector as long as gas consumption *during authentication* is reasonably low because (1)
    /// the blob storage capability bounds the number of unregistered blobs that can be submitted,
    /// and (2) if authentication succeeds then the gas for the blob is paid by the submitter.
    fn authenticate_unregistered<Accessor: ProvableStateReader<User, Spec = S>>(
        batch: &BatchFromUnregisteredSequencer,
        state: &mut Accessor,
    ) -> Result<AuthenticationOutput<S, Self::Decodable>, UnregisteredAuthenticationError>;

    /// Encode a standard transaction for the rollup with information describing how to authenticate it.
    fn add_standard_auth(tx: RawTx) -> Self::Input;

    /// Encode the input for the authenticator into a byte array.
    #[must_use]
    fn encode_authenticator_input(input: &Self::Input) -> FullyBakedTx {
        FullyBakedTx::new(borsh::to_vec(&input).unwrap())
    }

    /// Encode a standard transaction for the rollup with information describing how to authenticate it.
    #[must_use]
    fn encode_with_standard_auth(tx: RawTx) -> FullyBakedTx {
        Self::encode_authenticator_input(&Self::add_standard_auth(tx))
    }
}

/// See [`RollupAuthenticator`].
#[derive(std::fmt::Debug, Clone, borsh::BorshDeserialize, borsh::BorshSerialize)]
pub enum AuthenticatorInput {
    /// A rollup transaction.
    ///
    /// This is the only possible variant, but we keep this an `enum` instead of
    /// a `struct` to allow for future transaction types in a
    /// backwards-compatible way.
    Standard(RawTx),
}

/// Canonical implementation of [`TransactionAuthenticator`].
#[derive(Debug, PartialEq, Clone, Default)]
pub struct RollupAuthenticator<S, Rt>(PhantomData<(S, Rt)>);

impl<S, Rt> TransactionAuthenticator<S> for RollupAuthenticator<S, Rt>
where
    S: Spec,
    Rt: Runtime<S> + DispatchCall<Spec = S>,
{
    type Decodable = Rt::Decodable;
    type Input = AuthenticatorInput;

    #[cfg(feature = "native")]
    fn decode_serialized_tx(tx: &FullyBakedTx) -> Result<Self::Decodable, FatalError> {
        let AuthenticatorInput::Standard(tx) = borsh::from_slice(&tx.data)
            .map_err(|e| FatalError::DeserializationFailed(e.to_string()))?;

        capabilities::decode_sov_tx::<S, Rt>(&tx.data)
    }

    fn authenticate<Accessor: ProvableStateReader<sov_state::User, Spec = S>>(
        tx: &FullyBakedTx,
        pre_exec_ws: &mut Accessor,
    ) -> ::core::result::Result<
        capabilities::AuthenticationOutput<S, Self::Decodable>,
        capabilities::AuthenticationError,
    > {
        let AuthenticatorInput::Standard(input) = borsh::from_slice(&tx.data).map_err(|e| {
            capabilities::fatal_deserialization_error::<_, S, _>(&tx.data, e, pre_exec_ws)
        })?;

        crate::capabilities::authenticate::<_, S, Rt>(&input.data, &Rt::CHAIN_HASH, pre_exec_ws)
    }

    #[cfg(feature = "native")]
    fn compute_tx_hash(tx: &FullyBakedTx) -> anyhow::Result<TxHash> {
        let AuthenticatorInput::Standard(input) = borsh::from_slice(&tx.data)?;

        Ok(calculate_hash(
            &input.data,
            &mut crate::gas::UnlimitedGasMeter::<S>::default(),
        )?)
    }

    fn authenticate_unregistered<Accessor: ProvableStateReader<sov_state::User, Spec = S>>(
        batch: &BatchFromUnregisteredSequencer,
        pre_exec_ws: &mut Accessor,
    ) -> Result<AuthenticationOutput<S, Self::Decodable>, UnregisteredAuthenticationError> {
        crate::capabilities::authenticate::<_, S, Rt>(&batch.tx.data, &Rt::CHAIN_HASH, pre_exec_ws)
            .map_err(|e| match e {
                AuthenticationError::FatalError(err, hash) => {
                    UnregisteredAuthenticationError::FatalError(err, hash)
                }
                AuthenticationError::OutOfGas(err) => {
                    UnregisteredAuthenticationError::OutOfGas(err)
                }
            })
    }

    fn add_standard_auth(tx: RawTx) -> Self::Input {
        AuthenticatorInput::Standard(tx)
    }
}

/// Authorizes transactions to be executed.
pub trait TransactionAuthorizer<S: Spec> {
    /// Resolves the [`Context`] for a transaction.
    fn resolve_context(
        &mut self,
        auth_data: &AuthorizationData<S>,
        sequencer: &<<S as Spec>::Da as DaSpec>::Address,
        sequencer_rollup_address: S::Address,
        state: &mut impl StateAccessor,
    ) -> anyhow::Result<Context<S>>;

    /// Resolves the context for an unregistered transaction.
    fn resolve_unregistered_context(
        &mut self,
        auth_data: &AuthorizationData<S>,
        sequencer: &<<S as Spec>::Da as DaSpec>::Address,
        state: &mut impl StateAccessor,
    ) -> anyhow::Result<Context<S>>;

    /// Prevents duplicate transactions from running.
    fn check_uniqueness(
        &self,
        auth_data: &AuthorizationData<S>,
        context: &Context<S>,
        state: &mut impl StateAccessor,
    ) -> anyhow::Result<()>;

    /// Marks a transaction as having been executed, preventing it from executing again.
    fn mark_tx_attempted(
        &mut self,
        auth_data: &AuthorizationData<S>,
        sequencer: &<<S as Spec>::Da as DaSpec>::Address,
        state: &mut impl StateAccessor,
    ) -> anyhow::Result<()>;
}

/// Output of the authentication.
pub type AuthenticationOutput<S, Decodable> = (
    AuthenticatedTransactionAndRawHash<S>,
    AuthorizationData<S>,
    Decodable,
);

/// Error variants that can be raised as a [`AuthenticationError::FatalError`].
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Error)]
#[serde(rename_all = "snake_case")]
pub enum FatalError {
    /// Transaction deserialization failed.
    #[error("Transaction deserialization error: {0}")]
    DeserializationFailed(String),
    /// Signature verification failed.
    #[error("Signature verification failed: {0}")]
    SigVerificationFailed(String),
    /// The chain id was invalid
    #[error("Invalid chain id: expected {expected}, got {got}")]
    InvalidChainId {
        /// The expected chain id
        expected: u64,
        /// The actual chain id
        got: u64,
    },
    /// Transaction decoding failed.
    #[error("Transaction decoding error: {0}")]
    MessageDecodingFailed(String),
    /// A variant to capture any other fatal error.
    #[error("Other: {0}")]
    Other(String),
}

/// Authentication error type.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Error)]
#[serde(rename_all = "snake_case")]
pub enum AuthenticationError {
    /// The transaction authentication failed in a way that should have been detected by the sequencer before they accepted the transaction. The sequencer is slashed.
    #[error("Transaction authentication raised a fatal error, error: {0}, tx_hash {1}")]
    FatalError(FatalError, TxHash),
    /// The transaction authentication returned an error, but including it could have been an honest mistake. The sequencer should be charged enough to cover the cost of checking the transaction but not slashed.
    #[error("Transaction authentication ran out of gas: {0}.")]
    OutOfGas(
        /// The reason for the penalization.       
        String,
    ),
}

/// Authentication error relating to transactions submitted by an unregistered sequencer for the
/// purpose of direct sequencer registration.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Error)]
#[serde(rename_all = "snake_case")]
pub enum UnregisteredAuthenticationError {
    /// The transaction authentication failed in a way that is unrecoverable.
    #[error("Transaction authentication raised a fatal error, error: {0}")]
    FatalError(FatalError, TxHash),
    /// Transaction run out of gas
    #[error("Transaction ran out of gas, error: {0}")]
    OutOfGas(String),
}

/// The different types of data that can be used to verify transaction uniqueness
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum UniquenessData {
    /// Nonce-based uniqueness: an account's transactions must have a unique and consecutive nonces
    Nonce(u64),
    /// Generation-based uniqueness: the last `PAST_TRANSACTION_GENERATION` generations are cached.
    /// Transactions older than this buffer are invalid, transactions falling within it or with a
    /// higher generation are valid but must have a unique hash within their generation
    Generation(u64),
}

/// Data required to authorize a sov-transaction.
pub struct AuthorizationData<S: Spec> {
    /// The nonce of the transaction.
    pub uniqueness: UniquenessData,

    /// The hash of the transaction.
    pub tx_hash: TxHash,

    /// Credential identifier used to retrieve relevant rollup address.
    pub credential_id: CredentialId,

    /// Holds the original credentials to authenticate the transaction and
    /// provides information about which `Authenticator` was used to authenticate the transaction.
    pub credentials: Credentials,

    /// The default address.
    pub default_address: S::Address,
}

fn verify_and_decode_tx<S: Spec, D: DispatchCall<Spec = S>>(
    raw_tx_hash: TxHash,
    tx: Transaction<D, S>,
    chain_hash: &[u8; 32],
    meter: &mut impl GasMeter<Spec = S>,
) -> Result<AuthenticationOutput<S, D::Decodable>, AuthenticationError> {
    match &tx.versioned_tx {
        VersionedTx::V0(tx_v0) => {
            if tx_v0.details.chain_id != config_chain_id() {
                return Err(AuthenticationError::FatalError(
                    FatalError::InvalidChainId {
                        expected: config_chain_id(),
                        got: tx_v0.details.chain_id,
                    },
                    raw_tx_hash,
                ));
            }

            tx.verify(chain_hash, meter).map_err(|e| match e {
                TransactionVerificationError::BadSignature(_)
                | TransactionVerificationError::TransactionDeserializationError(_) => {
                    AuthenticationError::FatalError(
                        FatalError::SigVerificationFailed(e.to_string()),
                        raw_tx_hash,
                    )
                }
                TransactionVerificationError::GasError(_) => {
                    AuthenticationError::OutOfGas(e.to_string())
                }
            })?;

            let runtime_call = tx_v0.runtime_call.clone();
            let pub_key = tx_v0.pub_key.clone();
            let credential_id = metered_credential(&pub_key, meter)
                .map_err(|e| AuthenticationError::OutOfGas(e.to_string()))?;

            let default_address = credential_id.into();

            let credentials = Credentials::new(pub_key);
            let generation = tx_v0.generation;

            let tx_and_raw_hash = AuthenticatedTransactionAndRawHash {
                raw_tx_hash,
                authenticated_tx: AuthenticatedTransactionData(tx_v0.details.clone()),
            };

            Ok((
                tx_and_raw_hash,
                AuthorizationData {
                    uniqueness: UniquenessData::Generation(generation),
                    tx_hash: raw_tx_hash,
                    credential_id,
                    credentials,
                    default_address,
                },
                runtime_call,
            ))
        }
    }
}

/// Authenticate raw sov-transaction.
///
/// # Errors
/// Returns an error if gas runs out at any point, if deserialization or hashing fails, or if the
/// signature cannot be verified.
pub fn authenticate<
    Accessor: ProvableStateReader<User, Spec = S>,
    S: Spec,
    D: DispatchCall<Spec = S>,
>(
    mut raw_tx: &[u8],
    chain_hash: &[u8; 32],
    state: &mut Accessor,
) -> Result<AuthenticationOutput<S, D::Decodable>, AuthenticationError> {
    let raw_tx_hash = calculate_hash::<Accessor, S>(raw_tx, state)
        .map_err(|e| AuthenticationError::OutOfGas(e.to_string()))?;

    let tx =
        match <Transaction<D, S> as MeteredBorshDeserialize<S>>::deserialize(&mut raw_tx, state) {
            Ok(ok) => ok,

            Err(MeteredBorshDeserializeError::GasError(e)) => {
                return Err(AuthenticationError::OutOfGas(format!(
                    "Transaction deserialization run out of gas {e}, tx hash {raw_tx_hash}"
                )))
            }
            Err(MeteredBorshDeserializeError::IOError(e)) => {
                return Err(AuthenticationError::FatalError(
                    FatalError::DeserializationFailed(e.to_string()),
                    raw_tx_hash,
                ));
            }
        };

    verify_and_decode_tx::<S, D>(raw_tx_hash, tx, chain_hash, state)
}

/// Decode bytes as a Sovereign SDK transaction, returning the message and tx info.
#[cfg(feature = "native")]
pub fn decode_sov_tx<S: Spec, D: DispatchCall<Spec = S>>(
    mut raw_tx: &[u8],
) -> Result<D::Decodable, FatalError> {
    let tx = <Transaction<D, S> as MeteredBorshDeserialize<S>>::unmetered_deserialize(&mut raw_tx)
        .map_err(|e| FatalError::DeserializationFailed(e.to_string()))?;

    Ok(tx.call())
}

/// Calculates the hash of `data` and charges gas.
///
/// # Errors
/// Returns an error if the operation runs out of gas.
pub fn calculate_hash<G: GasMeter<Spec = S>, S: Spec>(
    data: &[u8],
    gas_meter: &mut G,
) -> Result<TxHash, GasMeteringError<S::Gas>> {
    let hash = MeteredHasher::<G, <S::CryptoSpec as CryptoSpec>::Hasher>::digest(data, gas_meter)
        .map(TxHash::new)?;

    Ok(hash)
}

/// Helper function to create a `FatalError::DeserializationFailed` authentication error.
pub fn fatal_deserialization_error<
    Accessor: ProvableStateReader<User, Spec = S>,
    S: Spec,
    E: ToString,
>(
    raw_tx: &[u8],
    err: E,
    pre_exec_working_set: &mut Accessor,
) -> AuthenticationError {
    let hash = match calculate_hash::<Accessor, S>(raw_tx, pre_exec_working_set) {
        Ok(hash) => hash,
        Err(err) => return AuthenticationError::OutOfGas(err.to_string()),
    };

    AuthenticationError::FatalError(FatalError::DeserializationFailed(err.to_string()), hash)
}
