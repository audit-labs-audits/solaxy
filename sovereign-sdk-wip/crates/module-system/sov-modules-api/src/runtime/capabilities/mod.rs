#![deny(missing_docs)]
//! The rollup capabilities module defines "capabilities" that rollup must
//! provide if they wish to use the standard app template.
//! If you don't want to provide these capabilities,
//! you can bypass the Sovereign module-system completely
//! and write a state transition function from scratch.
//! [See here for docs](https://github.com/Sovereign-Labs/sovereign-sdk/blob/nightly/examples/demo-stf/README.md)
pub mod auth;
mod batch_selector;
mod kernel;
mod proof;

#[cfg(feature = "native")]
use std::sync::Arc;

pub use auth::*;
pub use batch_selector::*;
pub use kernel::*;

mod gas;
pub use gas::*;
pub use proof::ProofProcessor;
mod sequencer;
pub use sequencer::*;
mod chain_state;
pub use chain_state::*;
use sov_rollup_interface::common::SlotNumber;

use crate::Spec;

/// Wrapper around an inner type that prevents accessing it.
pub struct Guard<T> {
    inner: T,
}

impl<T> Guard<T> {
    /// Create a new guarded instance.
    pub fn new(inner: T) -> Self {
        Self { inner }
    }
}

/// Indicates that a type provides the necessary capabilities for a runtime.
///
/// Capabilities are important core functionality like charging for gas, authorizing transactions,
/// and processing proofs. While overriding these capabilities is allowed and can be very useful, it is also dangerous -
/// using a non-standard implementation can cause subtle bugs. Some capabilities are coupled with one another - for example,
/// the sequencer registry sets asides funds that are used to reward the prover later on.
///
/// Take great care when overriding these capabilities. Always test your changes thoroughly, and be sure that you understand
/// the implications of your changes.
///
/// If you just want a sensible default implementation, use the `StandardCapabilities` struct provided by the SDK maintainers.
pub trait HasCapabilities<S: Spec> {
    /// The concrete implementation of the capabilities.
    type Capabilities<'a>: GasEnforcer<S>
        + SequencerAuthorization<S>
        + TransactionAuthorizer<S>
        + ProofProcessor<S>
        + SequencerRemuneration<S>
    where
        Self: 'a;

    /// Fetches the capabilities from the runtime.
    ///
    /// This method is only intended to be used internally on the [`HasCapabilities`] trait, if you
    /// need to access a capability do so with the constructor method.
    ///
    /// The returned struct is wrapped in a guard to prevent access from code outside of the trait.
    /// Without the guard it would be possible to implement an override for a capability and
    /// accidently use the default implementation leading subtle type mismatch bugs.
    ///
    /// For example, if I override [`HasCapabilities::gas_enforcer`] to return a different [`GasEnforcer`]
    /// implementation but then used `HasCapabilities::capabilities().try_reserve_gas` instead of
    /// `HasCapabilities::gas_enforcer().try_reserve_gas` I would use the default implementation instead of
    /// the override.
    fn capabilities(&mut self) -> Guard<Self::Capabilities<'_>>;

    /// Returns the [`GasEnforcer`] implementation on [`HasCapabilities::Capabilities`].
    ///
    /// This method can be overriden to provide a custom implementation.
    fn gas_enforcer(&mut self) -> impl GasEnforcer<S> {
        self.capabilities().inner
    }

    /// Returns the [`SequencerAuthorization`] implementation on [`HasCapabilities::Capabilities`].
    ///
    /// This method can be overriden to provide a custom implementation.
    fn sequencer_authorization(&mut self) -> impl SequencerAuthorization<S> {
        self.capabilities().inner
    }

    /// Returns the [`TransactionAuthorizer`] implementation on [`HasCapabilities::Capabilities`].
    ///
    /// This method can be overriden to provide a custom implementation.
    fn transaction_authorizer(&mut self) -> impl TransactionAuthorizer<S> {
        self.capabilities().inner
    }

    /// Returns the [`ProofProcessor`] implementation on [`HasCapabilities::Capabilities`].
    ///
    /// This method can be overriden to provide a custom implementation.
    fn proof_processor(&mut self) -> impl ProofProcessor<S> {
        self.capabilities().inner
    }

    /// Returns the [`SequencerRemuneration`] implementation on [`HasCapabilities::Capabilities`].
    ///
    /// This method can be overriden to provide a custom implementation.
    fn sequencer_remuneration(&mut self) -> impl SequencerRemuneration<S> {
        self.capabilities().inner
    }
}

/// Indicates that a type provides the necessary kernel capabilities for a runtime.
pub trait HasKernel<S: Spec>: Default + Send + Sync + 'static {
    /// The concrete implementation of the kernel.
    type Kernel<'a>: Kernel<S> + ChainState<Spec = S> + BlobSelector<Spec = S>
    where
        Self: 'a;

    /// Fetches the kernel modules from the runtime.
    ///
    /// This method is only intended to be used internally on the [`HasKernel`] trait, if you
    /// need to access a kernel capability do so with the constructor method.
    ///
    /// The returned struct is wrapped in a guard to prevent access from code outside of the trait.
    /// Without the guard it would be possible to implement an override for a kernel capability and
    /// accidently use the default implementation leading subtle type mismatch bugs.
    ///
    /// For example, if I override [`HasCapabilities::gas_enforcer`] to return a different [`GasEnforcer`]
    /// implementation but then used `HasCapabilities::capabilities().try_reserve_gas` instead of
    /// `HasCapabilities::gas_enforcer().try_reserve_gas` I would use the default implementation instead of
    /// the override.
    fn inner(&mut self) -> Guard<Self::Kernel<'_>>;

    /// Returns the [`Kernel`] implementation on [`HasKernel::Kernel`].
    fn kernel(&mut self) -> Self::Kernel<'_> {
        self.inner().inner
    }

    /// Returns the [`ChainState`] implementation on [`HasKernel::Kernel`].
    fn chain_state(&mut self) -> impl ChainState<Spec = S> {
        self.inner().inner
    }

    /// Returns the [`BlobSelector`] implementation on [`HasKernel::Kernel`].
    fn blob_selector(&mut self) -> impl BlobSelector<Spec = S> {
        self.inner().inner
    }

    /// Returns the [`KernelWithSlotMapping`] implementation on [`HasKernel::Kernel`].
    #[cfg(feature = "native")]
    fn kernel_with_slot_mapping(&self) -> Arc<dyn KernelWithSlotMapping<S>>;
}

#[cfg(feature = "test-utils")]
pub mod mocks {
    //! Mocks for the rollup capabilities module

    use sov_rollup_interface::common::{SlotNumber, VisibleSlotNumber};

    #[cfg(feature = "native")]
    use super::KernelWithSlotMapping;
    use super::{Kernel, RollupHeight, Spec};
    use crate::BootstrapWorkingSet;
    #[cfg(feature = "native")]
    use crate::GetGasPrice;

    /// A mock kernel for use in tests
    #[derive(Debug, Clone, Default)]
    pub struct MockKernel<S> {
        /// The current slot number
        pub true_slot_number: SlotNumber,
        /// The slot number at which transactions appear to be executing
        pub visible_slot_number: VisibleSlotNumber,
        /// The next sequence number to expect for preferred blobs.
        pub next_sequence_number: u64,
        phantom: core::marker::PhantomData<S>,
    }

    impl<S: Spec> MockKernel<S> {
        /// Create a new mock kernel with the given slot numbers
        pub fn new(true_slot_number: u64, visible_slot_number: u64) -> Self {
            Self {
                true_slot_number: SlotNumber::new_dangerous(true_slot_number),
                visible_slot_number: VisibleSlotNumber::new_dangerous(visible_slot_number),
                next_sequence_number: 0,
                phantom: core::marker::PhantomData,
            }
        }

        /// Simply increases all the heights by one
        pub fn increase_heights(&mut self) {
            self.true_slot_number.incr();
            self.visible_slot_number.incr();
        }
    }

    #[cfg(feature = "native")]
    impl<S: Spec> KernelWithSlotMapping<S> for MockKernel<S> {
        fn visible_slot_number_at(
            &self,
            true_slot_number: SlotNumber,
            _state: &mut crate::ApiStateAccessor<S>,
        ) -> Option<VisibleSlotNumber> {
            Some(true_slot_number.as_visible())
        }

        fn current_rollup_height(
            &self,
            _state: &mut crate::state::ApiStateAccessor<S>,
        ) -> RollupHeight {
            RollupHeight::new(self.visible_slot_number.get())
        }

        fn rollup_height_to_visible_slot_number(
            &self,
            height: super::RollupHeight,
            _state: &mut crate::state::ApiStateAccessor<S>,
        ) -> Option<VisibleSlotNumber> {
            Some(VisibleSlotNumber::new_dangerous(height.get()))
        }

        #[cfg(feature = "native")]
        fn true_slot_number_at_historical_height(
            &self,
            height: super::RollupHeight,
            _state: &mut crate::state::ApiStateAccessor<S>,
        ) -> Option<SlotNumber> {
            Some(SlotNumber::new_dangerous(height.get()))
        }

        fn base_fee_per_gas_at(
            &self,
            _height: super::RollupHeight,
            state: &mut crate::state::ApiStateAccessor<S>,
        ) -> Option<<<S as Spec>::Gas as crate::Gas>::Price> {
            Some(state.gas_price().clone())
        }

        fn true_slot_number_to_rollup_height(
            &self,
            _true_slot_number: SlotNumber,
            _state: &mut crate::state::ApiStateAccessor<S>,
        ) -> Option<RollupHeight> {
            Some(RollupHeight::new(self.visible_slot_number.get()))
        }
    }

    impl<S: Spec> Kernel<S> for MockKernel<S> {
        fn true_slot_number(&self, _ws: &mut BootstrapWorkingSet<'_, S>) -> SlotNumber {
            self.true_slot_number
        }
        fn next_visible_slot_number(
            &self,
            _ws: &mut BootstrapWorkingSet<'_, S>,
        ) -> VisibleSlotNumber {
            self.visible_slot_number
        }

        fn rollup_height(&self, _state: &mut BootstrapWorkingSet<'_, S>) -> super::RollupHeight {
            RollupHeight::new(self.visible_slot_number.get())
        }

        fn record_gas_usage(
            &mut self,
            _state: &mut crate::StateCheckpoint<S>,
            _final_gas_info: super::BlockGasInfo<<S as Spec>::Gas>,
            _rollup_height: RollupHeight,
        ) {
        }
    }
}

/// A rollup "block number". Rollup heights increase in order (1, 2, 3, ...),
/// regardless of what happens on the underlying DA layer.
#[derive(
    Clone,
    Copy,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    derive_more::Display,
    derive_more::FromStr,
    serde::Serialize,
    serde::Deserialize,
    borsh::BorshDeserialize,
    borsh::BorshSerialize,
)]
pub struct RollupHeight(u64);

impl std::fmt::Debug for RollupHeight {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.get())
    }
}

impl RollupHeight {
    /// The genesis rollup height.
    pub const GENESIS: Self = Self(0);

    /// The height of the first rollup block after genesis.
    pub const ONE: Self = Self(1);

    /// Create a new rollup height from a u64.
    pub fn new(height: u64) -> Self {
        Self(height)
    }

    /// Returns the inner value of a rollup height.
    pub fn get(&self) -> u64 {
        self.0
    }

    /// Increment a rollup height by one.
    pub fn incr(&mut self) {
        self.0 += 1;
    }

    /// See [`u64::checked_sub`]
    #[must_use]
    pub fn checked_sub(self, rhs: u64) -> Option<Self> {
        self.0.checked_sub(rhs).map(Self)
    }

    /// See [`u64::checked_add`]
    #[must_use]
    pub fn checked_add(self, rhs: Self) -> Option<Self> {
        self.0.checked_add(rhs.0).map(Self)
    }

    /// See [`u64::saturating_sub`]
    #[must_use]
    pub fn saturating_sub(self, rhs: u64) -> Self {
        Self(self.0.saturating_sub(rhs))
    }

    /// See [`u64::saturating_add`]
    #[must_use]
    pub fn saturating_add(self, rhs: u64) -> Self {
        Self(self.0.saturating_add(rhs))
    }

    /// Convert a rollup height to a slot number
    #[must_use]
    pub fn to_slot_number(&self) -> SlotNumber {
        SlotNumber::new_dangerous(self.0)
    }
}
