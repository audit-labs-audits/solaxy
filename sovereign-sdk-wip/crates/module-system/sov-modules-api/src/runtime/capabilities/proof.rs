use sov_rollup_interface::common::SlotNumber;
#[cfg(feature = "native")]
use sov_rollup_interface::optimistic::BondingProofService;
use sov_rollup_interface::optimistic::{SerializedAttestation, SerializedChallenge};
use sov_rollup_interface::stf::InvalidProofError;
use sov_rollup_interface::zk::aggregated_proof::{
    AggregatedProofPublicData, SerializedAggregatedProof,
};

#[cfg(feature = "native")]
use super::HasKernel;
#[cfg(feature = "native")]
use crate::rest::StateUpdateReceiver;
use crate::{GetGasPrice, SovAttestation, SovStateTransitionPublicData, Spec, Storage, TxState};

/// The `ProofProcessor` capability is responsible for processing proofs inside
/// the stf-blueprint.
///
/// ## Warning
/// The implementation of this trait is coupled with the implementation of the `GasEnforcer`, trait (and all of its dependencies),
/// since the `ProofProcessor` is responsible for charging disbursing gas fees to the prover.
pub trait ProofProcessor<S: Spec> {
    /// The service that generates bonding proofs for attesters.
    #[cfg(feature = "native")]
    type BondingProofService<K: HasKernel<S>>: BondingProofService;

    /// Creates a new [`BondingProofService`] service.
    #[cfg(feature = "native")]
    fn create_bonding_proof_service<K: HasKernel<S>>(
        &self,
        attester_address: <S as Spec>::Address,
        storage: StateUpdateReceiver<<S as Spec>::Storage>,
    ) -> Self::BondingProofService<K>;

    /// Called by the stf once the zk-proof is received.
    #[allow(clippy::type_complexity)]
    fn process_aggregated_proof<ST: TxState<S> + GetGasPrice<Spec = S>>(
        &mut self,
        proof: SerializedAggregatedProof,
        prover_address: &S::Address,
        state: &mut ST,
    ) -> Result<
        (
            AggregatedProofPublicData<S::Address, S::Da, <S::Storage as Storage>::Root>,
            SerializedAggregatedProof,
        ),
        InvalidProofError,
    >;

    /// Called by the stf once the attestation is received.
    fn process_attestation<ST: TxState<S> + GetGasPrice<Spec = S>>(
        &mut self,
        proof: SerializedAttestation,
        prover_address: &S::Address,
        state: &mut ST,
    ) -> Result<SovAttestation<S>, InvalidProofError>;

    /// Called by the stf once the challenge is received.
    fn process_challenge<ST: TxState<S> + GetGasPrice<Spec = S>>(
        &mut self,
        proof: SerializedChallenge,
        rollup_height: SlotNumber,
        prover_address: &S::Address,
        state: &mut ST,
    ) -> anyhow::Result<SovStateTransitionPublicData<S>, InvalidProofError>;
}
