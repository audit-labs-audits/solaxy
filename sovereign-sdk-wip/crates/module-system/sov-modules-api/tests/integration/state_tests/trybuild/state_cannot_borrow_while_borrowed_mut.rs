use sov_mock_zkvm::MockZkvm;
use sov_modules_api::capabilities::mocks::MockKernel;
use sov_modules_api::capabilities::RollupHeight;
use sov_modules_api::{StateCheckpoint, StateValue};
use sov_rollup_interface::execution_mode::Native;
use sov_state::{BorshCodec, Prefix};
use sov_test_utils::storage::SimpleStorageManager;
use sov_test_utils::MockDaSpec;
use unwrap_infallible::UnwrapInfallible;

type TestSpec = sov_modules_api::default_spec::DefaultSpec<MockDaSpec, MockZkvm, MockZkvm, Native>;

fn main() {
    let storage_manager = SimpleStorageManager::new();
    let storage = storage_manager.create_storage();
    let kernel = MockKernel::<TestSpec>::new(4, 1);
    let mut state = StateCheckpoint::new(storage, &kernel);

    let prefix = Prefix::new(b"test".to_vec());
    let mut value = crate::StateValue::<RollupHeight>::with_codec(prefix.clone(), BorshCodec);
    value.set(&RollupHeight::new(100), &mut state);
    let val = value.borrow_mut(&mut state).unwrap_infallible().unwrap();
    value.borrow(&mut state).unwrap_infallible();
    assert_eq!(*val, RollupHeight::new(100));
}
