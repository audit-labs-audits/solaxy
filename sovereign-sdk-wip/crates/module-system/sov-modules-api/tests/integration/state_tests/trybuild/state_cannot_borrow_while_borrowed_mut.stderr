error[E0502]: cannot borrow `value` as immutable because it is also borrowed as mutable
  --> tests/integration/state_tests/trybuild/state_cannot_borrow_while_borrowed_mut.rs:23:5
   |
22 |     let val = value.borrow_mut(&mut state).unwrap_infallible().unwrap();
   |               ----- mutable borrow occurs here
23 |     value.borrow(&mut state).unwrap_infallible();
   |     ^^^^^ immutable borrow occurs here
24 |     assert_eq!(*val, RollupHeight::new(100));
   |                 --- mutable borrow later used here
